libenv/sysinfo.c

```diff
@@ -2537,24 +2535,42 @@ static void Xen_Cpuid(uint32_t idx, uint32_t *eax, uint32_t *ebx, uint32_t *ecx,
 
 /******************************************************************/
 
-static int Xen_Hv_Check(void)
+static bool Xen_Hv_Check(void)
 {
-    uint32_t eax;
+    /* CPUID interface to Xen from arch-x86/cpuid.h:
+     * Leaf 1 (0x40000000)
+     * EAX: Largest Xen-information leaf. All leaves up to an including @EAX
+     *   are supported by the Xen host.
+     * EBX-EDX: "XenVMMXenVMM" signature, allowing positive identification
+     *   of a Xen host.
+     *
+     * Additional information can be found in the Hypervisor CPUID
+     * Interface Proposal (https://lkml.org/lkml/2008/10/1/246)
+     */
+
+    uint32_t eax, base;
     union
     {
         uint32_t u[3];
         char s[13];
     } sig = {{0}};
 
-    Xen_Cpuid(0x40000000, &eax, &sig.u[0], &sig.u[1], &sig.u[2]);
-
-    if (strcmp("XenVMMXenVMM", sig.s) || (eax < 0x40000002))
+    for (base = 0x40000000; base < 0x40010000; base += 0x100)
```

Comment:
> What is the logic in going from 0x40000000 to 0x40010000?

Because hypervisor CPUID leaves don't have a standard Xen can be found at the first unused 0x100 aligned offset between 0x40000000 and 0x40010000. While most of the time the vendor id will be found at address 0x40000000, to properly detect if Xen is the hypervisor it is necessary to scan this whole range.